### PR TITLE
CsvImport job

### PR DESCRIPTION
**Pull request by billyjacobson**
Adding a job that can be used to import a CSV

### CODE REVIEW

File: java/dataflow-connector-examples/src/main/java/com/google/cloud/bigtable/dataflow/example/CsvImport.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    static final DoFn<String[], Mutation> MUTATION_TRANSFORM = new DoFn<String[], Mutation>() {
+        @ProcessElement
+        public void processElement(DoFn<String[], Mutation>.ProcessContext c) throws Exception {
+            try {
+                String[] headers = c.getPipelineOptions().as(BigtableCsvOptions.class).getHeaders().split(",");
+
+                byte[] rowId = Bytes.toBytes(c.element()[0]);
+                Put row = new Put(rowId);
+                for (int i = 1; i < c.element().length; i++) {
+                    row.addColumn(FAMILY, Bytes.toBytes(headers[i]), Bytes.toBytes(c.element()[i]));
```
Comment (sduskis) [inline java/dataflow-connector-examples/src/main/java/com/google/cloud/bigtable/dataflow/example/CsvImport.java]: Please cache the `byte[]` of the headers to optimize performance of large jobs.

File: java/dataflow-connector-examples/src/main/java/com/google/cloud/bigtable/dataflow/example/CsvImport.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 2015 Google Inc. All Rights Reserved.
```
Comment (sduskis) [inline java/dataflow-connector-examples/src/main/java/com/google/cloud/bigtable/dataflow/example/CsvImport.java]: This should be `Copyright 2018 Google LLC.`

Comment (billyjacobson) [reply java/dataflow-connector-examples/src/main/java/com/google/cloud/bigtable/dataflow/example/CsvImport.java]: Done

File: java/dataflow-connector-examples/src/main/java/com/google/cloud/bigtable/dataflow/example/CsvImport.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        CloudBigtableTableConfiguration config =
+                new CloudBigtableTableConfiguration.Builder()
+                        .withProjectId(options.getBigtableProjectId())
+                        .withInstanceId(options.getBigtableInstanceId())
+                        .withTableId(options.getBigtableTableId())
+                        .build();
+
+        Pipeline p = Pipeline.create(options);
+
+        PCollection<String> lines = p.apply("ReadMyFile", TextIO.read().from(options.getInputFile()));
```
[the remark this answers is not in the document]
Comment (billyjacobson) [reply java/dataflow-connector-examples/src/main/java/com/google/cloud/bigtable/dataflow/example/CsvImport.java]: Done

File: java/dataflow-connector-examples/src/main/java/com/google/cloud/bigtable/dataflow/example/CsvImport.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+ */
+public class CsvImport {
+    private static final byte[] FAMILY = Bytes.toBytes("csv");
+    private static final Logger LOG = LoggerFactory.getLogger(ParseEventFn.class);
+
+    static final DoFn<String[], Mutation> MUTATION_TRANSFORM = new DoFn<String[], Mutation>() {
+        @ProcessElement
+        public void processElement(DoFn<String[], Mutation>.ProcessContext c) throws Exception {
+            try {
+                String[] headers = c.getPipelineOptions().as(BigtableCsvOptions.class).getHeaders().split(",");
```
[the remark this answers is not in the document]
Comment (billyjacobson) [reply java/dataflow-connector-examples/src/main/java/com/google/cloud/bigtable/dataflow/example/CsvImport.java]: I can't access the pipeline options in @setup right now: https://issues.apache.org/jira/browse/BEAM-950

File: java/dataflow-connector-examples/src/main/java/com/google/cloud/bigtable/dataflow/example/CsvImport.java (hunk 1 above)
Comment (billyjacobson) [reply java/dataflow-connector-examples/src/main/java/com/google/cloud/bigtable/dataflow/example/CsvImport.java]: done.

File: java/dataflow-connector-examples/src/main/java/com/google/cloud/bigtable/dataflow/example/CsvImport.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+ * --stagingLocation=<STAGING_LOCATION_IN_CLOUD_STORAGE> --runner=BlockingDataflowPipelineRunner In
+ * Eclipse, you can just modify the existing 'SERVICE' run configuration. The managed resource does
+ * not require the GOOGLE_APPLICATION_CREDENTIALS, since the pipeline will use the security
+ * configuration of the project specified by --project.
+ */
+public class CsvImport {
+    private static final byte[] FAMILY = Bytes.toBytes("csv");
+    private static final Logger LOG = LoggerFactory.getLogger(ParseEventFn.class);
+
+    static final DoFn<String[], Mutation> MUTATION_TRANSFORM = new DoFn<String[], Mutation>() {
```
[the remark this answers is not in the document]
Comment (billyjacobson) [reply java/dataflow-connector-examples/src/main/java/com/google/cloud/bigtable/dataflow/example/CsvImport.java]: done.

**Comment by billyjacobson** [comment]
@sduskis @jabubake can you review/approve? Thanks

File: java/dataflow-connector-examples/src/main/java/com/google/cloud/bigtable/dataflow/example/CsvImport.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+  private static final byte[] FAMILY = Bytes.toBytes("csv");
+  private static final Logger LOG = LoggerFactory.getLogger(CsvImport.class);
+
+  static final DoFn<String, Mutation> MUTATION_TRANSFORM = new DoFn<String, Mutation>() {
+    @ProcessElement
+    public void processElement(DoFn<String, Mutation>.ProcessContext c) throws Exception {
+      try {
+        String[] headers = c.getPipelineOptions().as(BigtableCsvOptions.class).getHeaders()
+            .split(",");
+        String[] values = c.element().split(",");
```
Comment (sduskis) [inline java/dataflow-connector-examples/src/main/java/com/google/cloud/bigtable/dataflow/example/CsvImport.java]: Please add a `Preconditions` to make sure that the sizes of headers and values match.

File: java/dataflow-connector-examples/src/main/java/com/google/cloud/bigtable/dataflow/example/CsvImport.java [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+        byte[] rowkey = Bytes.toBytes(values[0]);
+        byte[][] headerBytes = new byte[headers.length][];
+        for (int i = 0; i < headers.length; i++) {
+          headerBytes[i] = Bytes.toBytes(headers[i]);
+        }
+
+        Put row = new Put(rowkey);
+        for (int i = 1; i < values.length; i++) {
+          row.addColumn(FAMILY, headerBytes[i], Bytes.toBytes(values[i]));
```
Comment (sduskis) [inline java/dataflow-connector-examples/src/main/java/com/google/cloud/bigtable/dataflow/example/CsvImport.java]: Please explicitly add the client-side timestamp in `addColumn`.

File: java/dataflow-connector-examples/src/main/java/com/google/cloud/bigtable/dataflow/example/CsvImport.java [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        }
+        c.output(row);
+      } catch (Exception e) {
+        LOG.error("Failed to process input {}", c.element(), e);
```
Comment (sduskis) [inline java/dataflow-connector-examples/src/main/java/com/google/cloud/bigtable/dataflow/example/CsvImport.java]: Please don't swallow the exception.  This exception needs to be bubbled up to Dataflow.

**Pull request merged**